File: jsfring-webservice/src/main/java/fr/pinguet62/jsfring/webservice/controller/RightController.java
package fr.pinguet62.jsfring.webservice.controller;

import fr.pinguet62.jsfring.service.RightService;
import fr.pinguet62.jsfring.webservice.converter.RightMapper;
import fr.pinguet62.jsfring.webservice.dto.RightDto;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import static fr.pinguet62.jsfring.webservice.controller.RightController.PATH;
import static java.util.Objects.requireNonNull;

@RestController
@RequestMapping(PATH)
public final class RightController {

    public static final String PATH = "/right";

    private final RightService rightService;

    private final RightMapper converter;

    public RightController(RightService rightService, RightMapper converter) {
        this.rightService = requireNonNull(rightService);
        this.converter = requireNonNull(converter);
    }

    @GetMapping("/{code}")
    public Mono<RightDto> get(@PathVariable String code) {
        return rightService
                .get(code)
                .map(converter::toDto);
    }

    @GetMapping
    public Flux<RightDto> list() {
        return rightService
                .getAll()
                .map(converter::toDto);
    }

    @PostMapping
    public void update(@RequestBody RightDto rightDto) {
        rightService
                .get(rightDto.getCode())
                .doOnNext(right -> converter.updateFromDto(right, rightDto))
                .doOnNext(rightService::update);
    }

}

File: jsfring-webservice/src/main/kotlin/fr/pinguet62/jsfring/webservice/controller/RightController.kt
package fr.pinguet62.jsfring.webservice.controller

import fr.pinguet62.jsfring.service.RightService
import fr.pinguet62.jsfring.webservice.controller.RightController.Companion.PATH
import fr.pinguet62.jsfring.webservice.converter.RightMapper
import fr.pinguet62.jsfring.webservice.dto.RightDto
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(PATH)
class RightController(
        val rightService: RightService,
        val converter: RightMapper
) {

    companion object {
        const val PATH = "/right"
    }

    @GetMapping("/{code}")
    fun get(@PathVariable code: String) =
            rightService
                    .get(code)
                    .map { converter.toDto(it) }

    @GetMapping
    fun list() =
            rightService
                    .getAll()
                    .map { converter.toDto(it) }

    @PostMapping
    fun update(@RequestBody rightDto: RightDto) {
        rightService
                .get(rightDto.code)
                .doOnNext { converter.updateFromDto(it, rightDto) }
                .doOnNext { rightService.update(it) }
    }

}